keps/140-scenario-based-simulation/README.md

```diff
@@ -0,0 +1,397 @@
+  PreemptedBy         *string
+  // CreatedAt indicates when the Pod was created.
+  CreatedAt           FakeTime
+  // BoundAt indicates when the Pod was scheduled.
+  // This field may be nil if this Pod has not been scheduled.
+  BoundAt             *FakeTime
+  // PreemptedAt indicates when the Pod was preempted.
+  // This field may be nil if this Pod has not been preempted.
+  PreemptedAt         *FakeTime
+  // ScheduleResult has the results of all scheduling for the Pod.
+  ScheduleResult      []ScenarioPodScheduleResult
+}
+
+type ScenarioPodScheduleResult struct {
+  // Time indicates the time scheduling at which the scheduling is performed.
+  Time                *FakeTime
+  // AllCandidateNodes indicates all candidate Nodes before Filter.
+  AllCandidateNodes   []string
+  // AllFilteredNodes indicates all candidate Nodes after Filter.
+  AllFilteredNodes    []string
+  // PluginResults has each plugin’s result.
+  PluginResults       ScenarioPluginsResults
+}
+
+type (
+  NodeName   string
+  PluginName string
+)
+
+type ScenarioPluginsResults struct {
+  // Filter has each filter plugin’s result.
+  Filter            map[NodeName]map[PluginName]string
+  // Score has each score plugin’s score.
+  Score             map[NodeName]map[PluginName]ScenarioPluginsScoreResult
+}
+
+type ScenarioPluginsScoreResult struct {
+  // RawScore has the score from Score method of Score plugins.
+  RawScore             int64
+  // NormalizedScore has the score calculated by NormalizeScore method of Score plugins.
+  NormalizedScore      int64
+  // FinalScore has score plugin’s final score calculated by normalizing with NormalizedScore and applied Score plugin weight. 
+  FinalScore           int64
+}
+```
+
+#### The concept "FakeTime"
+
+FakeTime is the concept to represent time in simulation, not a difficult one. 
+- In FakeTime, time is simply represented by numbers. like 1, 2, 3…
+- In FakeTime, time is moved to next time **when it can no longer schedule any more Pods**.
+
+The following shows what happens at a single time in FakeTime:
+
+1. run all operations defined for that time
+2. scheduler starts scheduling
+3. scheduler stops scheduling when it can no longer schedule any more Pods.
+4. update status.scenarioResults.
+5. move to next time.
+
+##### Why scheduler needs to restarts/stops scheduling loop?
+
+To ensure that the results of the simulation do not vary significantly from run to run. 
+
+When users define multiple operations for the same time, users expect them to run concurrently, but in practice it is difficult to run them at the same time. This is because the scheduler is constantly attempting to schedule. 
+For example, suppose a user has defined a scenario to create 1000 Nodes at the same time. Since it is strictly impossible to create 1000 Nodes at the same time, pending Pods will be scheduled to the Nodes created first. And depending on what order the Nodes were created, the results of the simulation may change.
```

Comment:
```suggestion
For example, suppose a user has defined a scenario to create 1000 Nodes in the same `FakeTime`. Since it is strictly impossible to create 1000 Nodes at the same time, pending Pods will be scheduled to the Nodes created first. And depending on what order the Nodes were created, the results of the simulation may change.
```